### ISSUE TITLE
[UX] Add a "Remove block" link in the block contextual links (with modal confirmation).

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#1725]
Part of #1723 

A bit of a WTF this one when a user wants to remove a block from the front-end...
1. Hover over block in question and click on the contextual links cog.
2. Click on the "configure block" contextual link and be brought to the respective layout edit form with the "Configure block" dialog for that block already open for them.
3. Wait! ...there's no "remove" button available -> Close the dialog.
4. Scroll to the part of the layout form that holds the block (#1724)...
5. Hit the down arrow on the "configure" drop-button...
6. Hit the "Remove" option.
7. Save changes -> back to the front-end with the block removed.

We could add a "Remove block" button in the dialog so that the user could have this action available to them from step 2 above. Then the same process would be:
1. Hover over block in question and click on the contextual links cog.
2. Click on the "configure block" contextual link and be brought to the respective layout edit form with the "Configure block" dialog for that block already open for them.
3. Click on the "Remove block" button...
4. Save changes -> back to the front-end with the block removed.

3 less clicks to accomplish the same task! Much better. UX+ ...but, if we had an actual "Remove block" contextual link additionally to the "Configure block" one, then this would be:
1. Hover over block in question and click on the contextual links cog.
2. Click on the "remove block" contextual link.
3. Confirmation dialog pops-up -> click "Remove" button to confirm -> Block removed without ever leaving the front-end.

1 less click then the previous method **and** a couple less page loads. UX++

---
PR by @docwilmot https://github.com/backdrop/backdrop/pull/1615

**Comment by quicksketch** [comment]
> ...but, if we had an actual "Remove block" contextual link additionally to the "Configure block" one, then this would be:

I like the idea of a contextual link for "Remove block" the best, but I think the user should still be taken to the back-end to do the confirmation and saving. If you don't go to the back end, you have no idea which layout you're affecting. Are you removing this block from just this page or all pages? You won't know unless you see which layout configuration you are editing/saving. So I'd like to see:

1. Hover over block in question and click on the contextual links cog.
2. Click on the "remove block" contextual link.
3. Taken to the back-end. Confirmation dialog pops-up -> click "Remove" button to confirm.
4. Save changes -> back to the front-end with the block removed.

Basically the same as your first suggestion but without the mix of terminology ("click configure" then "remove" -- instead "click remove" then "remove").

**Comment by docwilmot** [comment]
I'm leaning towards the third option personally.A remove/delete is a decisive thing, you dont expect options, or settings, you just want it to go, except for a confirmation dialog. To go to the backend here just seems unnatural. 

I propose:
1. Hover over block in question and click on the contextual links cog.
2. Click on the "remove block" contextual link.
3. Confirmation dialog pops-up which offers checkboxes for layouts where the block is active along with a "remove all" option.
4. Click Remove button, done.

**Comment by klonos** [comment]
> Confirmation dialog pops-up which offers checkboxes for layouts where the block is active along with a "remove all" option.

👍 ...million!

**Comment by docwilmot** [comment]
I just had a look at this and one small issue is that a dialog on the front end uses the site theme. 😄 
Was a bit of an unexpected shock.
![modal](https://cloud.githubusercontent.com/assets/1042283/19414340/1fe3f68a-930f-11e6-8e64-77c9ee78edbd.PNG)

**Comment by olafgrabienski** [comment]
@docwilmot Is there a PR to have a look at it? Looking at the screenshot, the button text "Save configuration" seems too neutral to me. Should be "Delete [block name] block" or so, in my opinion.

**Comment by docwilmot** [comment]
@olafgrabienski this image was just to show the theme.

**Comment by olafgrabienski** [comment]
I see! Thanks for the clarification.

**Comment by docwilmot** [comment]
First pass up.

**Comment by olafgrabienski** [comment]
I tested the PR (without actually deleting a block). That's what it looks like:

![screen-backdrop-delete-block-dialog](https://cloud.githubusercontent.com/assets/14188307/19433056/71fbdd7e-945f-11e6-8bdd-e844e0a3146f.png)

A nice UX improvement in my opinion. Two observations:

- The confirmation dialog is missing the "remove all" option mentioned in https://github.com/backdrop/backdrop-issues/issues/1725#issuecomment-244800500.
- Additionally to the **X** button, the dialog could need a "Cancel" link.

**Comment by olafgrabienski** [comment]
... another observation: When I hit "Delete" without having checked one of the options, I get a "HTTP ERROR 500" in the browser.

**Comment by klonos** [comment]
I agree with @olafgrabienski on the addition of a cancel link. A coupe of other minor improvements to consider:

a) Perhaps use the verb "remove" instead of "delete". I think it describes more accurately the action being performed and it would make it consistent with the respective drop-button action that's available in the admin UI.

b) Perhaps also have the checkbox for the current layout (the one the delete action was initiated from) checked by default. That would require 1 less click from the user (less clicks = better UX).

c) As an alternative to b above, we could exclude the current layout from the checkbox list and have the dialog message say:

  "The `%block_name%` block will be removed from the %current_layout_region% region of the `%current_layout_name%` layout.".

  If the block is used in other layouts/regions besides the current one, then also have the following added to the dialog text:

  "You can optionally select to remove it from the following layouts/regions where it is also present:  `checkbox list of layouts/regions`"

**Comment by docwilmot** [comment]
> The confirmation dialog is missing the "remove all" option mentioned in #1725 (comment).

I thought that since the removal of a block is a destructive process, it was reasonable to have users manually select for deletion rather than a "remove all" which could leave unfortunate accidentsal deletes.

> Additionally to the X button, the dialog could need a "Cancel" link.

Good option, will add.

> When I hit "Delete" without having checked one of the options, I get a "HTTP ERROR 500" in the browser.
 
Oops, sorry, will fix.

> Perhaps use the verb "remove" instead of "delete"

Reasonable.

> Also append the respective region names next to the layouts.

Don't think this adds much benefit TBH, but await consensus.

> Perhaps also have the checkbox for the current layout checked by default.
> As an alternative ... we could exclude the current layout from the checkbox list ... if the block is used in other layouts/regions besides the current one, then also have the following added to the dialog text

I love both suggestions; I'm not sure which I prefer. I would appreciate some more feedback on this.

**Comment by olafgrabienski** [comment]
> I thought that since the removal of a block is a destructive process, it was reasonable to have users manually select for deletion rather than a "remove all" ...

Good decision, in my opinion.

> checkbox for the current layout checked by default (vs. alternative)

I'd prefer the first suggestion, seems less complicated to me.

**Comment by docwilmot** [comment]
Updated, for review
Didnt add a remove link because whatever code responsible for forcing links into the actions area on dialogs doesnt seem to work. And there's always the `[X]`

**Comment by olafgrabienski** [comment]
> Didnt add a remove link because whatever code responsible for forcing links into the actions area on dialogs doesnt seem to work. And there's always the [X]

Reasonable, in my opinion!

I've had a look on the sandbox site at the following changes:

- Use the verb "Remove" instead of "Delete" => done
- Hit "Remove" without having checked one of the options => works, you get a relevant message
- Append the respective region names next to the layouts => done

I discovered two possible issues as well:

(1) When you have two same blocks on a layout, putting for example another "Powered by Backdrop CMS" in the sidebar of the Default layout, and removing it then via the Contextual link, the "Powered by"-Block in the Footer gets removed as well. May be an edge case, but in such cases it could confuse users a lot.

(2) Minor styling issue (tested in Chrome on Mac): The position of the X-Button doesn't align well, see screenshot:

![screen-backdrop-pr-1615-close-button-a](https://cloud.githubusercontent.com/assets/14188307/25329945/1b61b754-28de-11e7-9d26-cb5da7f3a42d.png)

Relevant code:

```
.ui-dialog .ui-dialog-titlebar-close {
  top: calc(-0.125em);
  right: calc(-0.125em);
  height: calc(100% + 0.125em);
}
```

`0.125em` seems to equal `2px`, I guess `3px` would fit better.

**Comment by docwilmot** [comment]
> When you have two same blocks on a layout...gets removed as well

Well I should have seen that coming. Will alter the code to remove blocks per region rather than per layout. 

The CSS thing is a Basis theme problem. Recommend open a new issue please.

**Comment by klonos** [comment]
>- Hit "Remove" without having checked one of the options => works, you get a relevant message

I'd prefer it if there was a "no selection made" text somewhere within that dialog and/or the remove button was disabled/grayed-out/hidden until a selection is made.

>- Append the respective region names next to the layouts => done

Nope. Not done.

> The CSS thing is a Basis theme problem. Recommend open a new issue please.

I did open #2653, but I cannot reproduce this styling issue with the dialogs on a vanilla 1.x installation. So, I'm guessing that either this PR is introducing/removing something, or the fact that dialogs are have not been tested in the front-end has something to do with this.

**Comment by olafgrabienski** [comment]
> Append the respective region names next to the layouts => done

> Nope. Not done.

I see, that was a misunderstanding on my part.

**Comment by serundeputy** [comment]
@docwilmot this looks great!

I'm fine with merging as is great feature.

I took a look at adding a cancel button; something like this:

```php
  $form['actions'] = array('#type' => 'actions');
  $form['actions']['submit'] = array(
    '#type' => 'submit',
    '#value' => t('Remove'),
    '#attributes' => array(
      'class' => array('button-danger',),
    ),
  );
  $form['actions']['cancel'] = array(
    '#type' => 'submit',
    '#value' => t('Cancel'),
    '#attributes' => array(
      'class' => array('button-secondary',),
    ),
    '#submit' => array(
      'layout_remove_block_canel_action'
    ),
  );

  return $form;
}

function layout_remove_block_canel_action(&$form, &$form_state) {
  // Just call this to close the dialogue.
}
```

I'm not sure about having that empty callback `layout_remove_block_canel_action` nor did I spend a lot of time thinking about a proper name.

What do you think @docwilmot want to add a cancel button and `button-danger` class or merge as is?

thanks,

**Comment by docwilmot** [comment]
@serundeputy I wasnt sure a canel *button* would do; a cancel *link* was what was requested. If yuo feel a button would do, I'll add that instead. 

And dont merge yet. I'm not happy with the fact that every instance of a block delta disappears once you remove one. I imagine if a person added two of a block by accident it would be annoying to try and delete one and have both disappear.

Working now. PR soon.

**Comment by docwilmot** [comment]
Pushed new. Can now delete individual blocks at will. Review welcome please.
Also added cancel button and also not sure about the empty function.

**Comment by Dinornis** [comment]
Just tested the PR and everything seems to be working just fine, well done everyone.
This is a great UX improvement.

**Comment by olafgrabienski** [comment]
Had a look at the sandbox site as well, deleting of individual blocks works now, great!

Re *Cancel* link or button: If we take the button which is okay in my opinion (don't know the Backdrop UX guidelines), I'd make it gray instead of blue. Two buttons of the same color look strange to me, see screenshot:

![screen-backdrop-pr-1615-cancel-button](https://cloud.githubusercontent.com/assets/14188307/25426815/a7af9d46-2a70-11e7-8afa-232fa236ca25.png)

**Comment by klonos** [comment]
> Two buttons of the same color look strange to me, see screenshot:

The remove button should be a danger one as @serundeputy suggested. Removing a block from the manage blocks UI gives you a second chance to cancel saving the layout (although I'd love it if there was some sort of undo). With this feature here though the blocks are removed straight away, so we should have that fact be reflected by a red button.

**Comment by Dinornis** [comment]
Why is the _Remove block_ link `/admin/structure/layouts/manage/home/contextual-remove/` load the _Basis_ style sheet at `/core/themes/basis/css/skin.css` instead of the _Seven_ style sheet at `/core/themes/seven/css/style.css` ?

The _Configure block_ link pointing to _/admin/structure/layouts/manage/home/configure-redirect/_ uses the correct style sheet.

**Comment by docwilmot** [comment]
> The remove button should be a danger one as @serundeputy suggested

True, missed that.
Note that there seems to be some class warfare going on with buttons in Basis: the `.button-secondary` class should be white with a gray border and grey text, but that CSS is being overridden by a `.ui-state-default, .ui-widget-content .ui-state-default` class in Basis which makes it blue again.

> Why is the Remove block link /admin/structure/layouts/manage/home/contextual-remove/ load the Basis style sheet

Because the dialog is on the front end and using Basis theme. I actually wonder if there may be a way to override the theme a dialog uses, but I dont know how to do that. maybe @quicksketch might.

**Comment by Dinornis** [comment]
I see, thanks for explaining that.
Maybe the function  `Use the administration theme when editing or creating content` could be amended to also include admin dialogs like this one.

**Comment by quicksketch** [comment]
> Note that there seems to be some class warfare going on with buttons in Basis: the .button-secondary class should be white with a gray border and grey text, but that CSS is being overridden by a .ui-state-default, .ui-widget-content .ui-state-default class in Basis which makes it blue again.

I just ran into this in the Views UI refresh (https://github.com/backdrop/backdrop-issues/issues/930). Very annoying. I'm not sure what the solution is here.

> I actually wonder if there may be a way to override the theme a dialog uses, but I don't know how to do that.

I made a separate issue for this: [#2659: Allow admin theme CSS for dialogs to be used on front-end](https://github.com/backdrop/backdrop-issues/issues/2659)

I'm still opposed to adding this highly-administrative functionality to the front-end of the site. I find the mixing of functionality within the front-end to be disturbing. It's a slippery slope when administrative functionality is moved into the front-end.

I can concede on my personal opinion though. There's no arguing that it's faster and easier using the dialog. I'm not sure I can make a solid case when weighed [our values](https://backdropcms.org/philosophy) either.

**Comment by quicksketch** [comment]
(My complaint aside, this functionality is awesome)

**Comment by quicksketch** [comment]
I posted a review to https://github.com/backdrop/backdrop/pull/1615#pullrequestreview-35487673. Besides the (minor) requested changes, this PR needs tests to confirm that the multiple delete ability works.

Great job making this functionality work without JS as well. I can middle-click on the "Remove block" link to open the confirm form in its own page. That should make this fairly easy test.

**Comment by jenlampton** [comment]
Oh man, this is quite the issue. I'm sorry I missed out on the conversation until now!

I feel very strongly that making this kind of change would create a considerably more confusing user experience. This is why:

> ... you have no idea which layout you're affecting. Are you removing this block from just this page or all pages? You won't know unless you see which layout configuration you are editing/saving.

I'd go even further and say that a user who clicked a contextual link from the front-end of the site still wouldn't know how many *pages* would be affected from the layout configuration "manage blocks" interface, as the scope of the layout is not defined there. 

This is the same objection I have with the inline editor: From the front end, users are missing the context of what's being edited. With a CMS there are too many patterns in play. How many places does that thing you are editing or removing appear? It's impossible to tell when you are only looking at one of them.

Yes, this kind of change might make it more convenient for some of the experts, but we'll be hurting the beginners in doing so. I could see myself getting tripped up by something like this!

**Comment by jenlampton** [comment]
more thoughts...  I suppose we are already in this boat by having the `configure` contextual link on blocks. Configuring a block -- thinking you are affecting only one page -- and accidentally changing that block's settings on most pages, is nearly as bad.

**Comment by jenlampton** [comment]
so many thoughts...

There is one feature from Panels (Drupal 7) that we are missing in Backdrop that might serve the same goal, in a slightly less dangerous way: A contextual link on the layout itself. 

I find it helpful to be able to "configure" the layout I'm looking at. It's essentially a shortcut from the front end to the back end, with the current layout already "selected". It doesn't provide any immediate actions (like deleting/configuring a block) but it does help me identify which layout is in use on that page I was just looking at.

Maybe adding more tools like this that help the site-architects by taking them to the places they need to be could be a compromise?

**Comment by klonos** [comment]
I still think this is a great UX improvement (the target audience being site builders).

> I suppose we are already in this boat by having the `configure` contextual link on blocks. Configuring a block -- thinking you are affecting only one page -- and accidentally changing that block's settings on most pages, is nearly as bad as removing it.

Exactly. The disconnection between front-end and backend when it comes to layouts is there as it is now; it's not a problem that this feature introduces. When a user is removing a block from the manage blocks UI, there is no *direct* indication of how many things they are affecting; the user simply needs to be aware of the overall configuration of the layout they are editing (context, visibility conditions other blocks configuration etc) and do all their edits based on that knowledge. I think that by listing the layouts and regions from which the block is to be removed on that modal works great in communicating what is going to happen. Since this is a destructive action, making the remove button a red, danger one would also helt mitigate the risk of accidental deletions.

**Comment by klonos** [comment]
...having said all the above, I have previously stated that ideally there would be a) configuration revisions so that people can undo/revert accidental actions, and b) trash bin for accidentally deleted users/nodes too (#2498).

**Comment by jenlampton** [comment]
...and I still think this would be a UX regression. We'd be taking one little UX problem we already have, and we'd be using that as a 'model' (actually making the problem worse, as we do so).

I know the expert-site builders (like @klonos) would appreciate having like a tool like this -- but those are the people who are the most likely to know how to search contrib to find one. Putting something this dangerous in core - and into the hands of the beginners and first-time evaluators seems like asking for trouble.

At the least I'd like to see this feature postponed until 1.8 - so more people have a chance to weigh in. Best case scenario we could develop this tool in contrib and watch it's usage statistics.

**Comment by klonos** [comment]
Fair enough about postponing till we get more feedback, but do you also think that we should remove the "configure block" link (which this issue here was to compliment)?

**Comment by klonos** [comment]
...or perhaps replace the "configure block" link with a more general "edit layout" link?

**Comment by jenlampton** [comment]
> do you also think that we should remove the "configure block" link

Not in 1.7, and not until we get more feedback on that too. ;) In general, that kind of tool makes me uneasy and I would like to see it removed (or improved) -- but maybe only after we have better tools for the advanced editors in place. I wouldn't want to remove something people use without a better replacement.

> ...or perhaps replace the "configure block" link with a more general "edit layout" link?

Yes.

**Comment by quicksketch** [comment]
Is a valid compromise to just not use the modal? Take the user to the back-end and show the exact same form as a page?

**Comment by docwilmot** [comment]
I think that would obfuscate the process rather than help TBH.

My opinion on this whole issue is that removal of a block from a layout is a destructive process in every situation, Drupal, Layouts, modals, whatever. You click remove and confirm and its gone.

I hear the concern but I dont agree that it is any more significant with this method. At least with this PR you are able to visualize exactly which block you are removing, rather than a representation of that block in Layout UI (less clear) or on a dedicated page (much less clear). In none of these situations is it obvious which pages you are deleting the block from. Again though, this PR shows you clearly which block, which layout, which paths, which regions, all clearly in a confirmation form. In all situations you click that confirm/save button on your own choice.

I do not see this as a significant difference in risk from the Layout UI.

Having said, if there is this much uncertainty so close to freeze, I'm happy to see it postponed (again). No worries.

**Comment by jenlampton** [comment]
> Is a valid compromise to just not use the modal?

I don't know, if we decide to remove the problematic contextual links from blocks entirely, adding another one now seems silly. I think there's still too much TBD.

**Comment by jenlampton** [comment]
> Again though, this PR shows you clearly which block, which layout, which paths, which regions, all clearly in a confirmation form.

Aha, it sounds like this confirm form is what we need to fix the current UX problem everywhere settings are changed or blocks are removed.

**Comment by jenlampton** [comment]
I think we should push this feature to 1.8 so that we have some more time to think about it. 

It's a slippery slope from something like this to editing/configuring all back-end-things (layouts, views, fields, etc) on the front-end too.  Maybe that's what we want? Let's get some more feedback :)

**Comment by klonos** [comment]
> Aha, it sounds like this confirm form is what we need to fix the current UX problem everywhere settings are changed or blocks are removed.

I think I've previously filed an issue about visually indicating that blocks are to be removed by some sort of red text and matching styling. [goes searching in the queue...]

**Comment by klonos** [comment]
...Aha!!: #2430

**Comment by jenlampton** [comment]
We are 2 weeks from feature freeze for 1.8.  Has anyone had time to think about this front-end vs back-end configure form problem, and do we have any new or differing opinions?

**Comment by klonos** [comment]
I personally still find this change to be a UX improvement.

There are a couple of issues with the styling in this PR:

![screen shot 2017-09-03 at 3 30 56 pm](https://user-images.githubusercontent.com/2423362/30000832-143395f0-90bd-11e7-8d83-0c0f75512227.png)

1. The z-index of the modal needs to be adjusted to be above the main nav menu.
2. The height and positioning of the close button is a few pixels off.

**Comment by jenlampton** [comment]
With 1 day until code freeze, this change is not likely to make it into 1.8. We haven't had very much discussion about the precedent making a UX change as major would set, and what would mean for the whole direction of Backdrop administration. We don't have time to make have that discussion in 1 day, along with everything else we'd like to have done. Bumping milestone.

**Comment by docwilmot** [comment]
Moved to contrib. https://github.com/backdrop-contrib/block_remove

**Comment by klonos** [comment]
Thanks @docwilmot for putting that into contrib, but I still think that there is merit in addressing this UX problem in core. If the blocker here is the whole "should not be editing backend things in the frontend", then at the very minimum, we should consider adding "remove" links/buttons in the block configuration dialogs. I've created a separate issue to consider that here: #4449
